pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,248 @@
+package routemetrics
+
+import (
+	"context"
+	"fmt"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	routev1 "github.com/openshift/api/route/v1"
+)
+
+const (
+	controllerName = "route_metrics_controller"
+)
+
+var (
+	log = logf.Logger.WithName(controllerName)
+)
+
+// New creates the route metrics controller. This is the controller
+// that handles all the logic for gathering and exporting
+// metrics related to route resources.
+func New(mgr manager.Manager) (controller.Controller, error) {
+	// Create a new cluster with a new cache to watch on Route objects from every namespace.
+	newCluster, err := cluster.New(mgr.GetConfig(), func(o *cluster.Options) {
+		o.Scheme = mgr.GetScheme()
+	})
+	if err != nil {
+		return nil, err
+	}
+	// Add the cluster to the manager so that the cluster is started along with the other runnables.
+	mgr.Add(newCluster)
+	reconciler := &reconciler{
+		client: mgr.GetClient(),
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	// add watch for changes in IngressController
+	if err := c.Watch(&source.Kind{Type: &operatorv1.IngressController{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+	// add watch for changes in Route
+	if err := c.Watch(source.NewKindWithCache(&routev1.Route{}, newCluster.GetCache()),
+		handler.EnqueueRequestsFromMapFunc(func(client.Object) []reconcile.Request { return nil }),
```

Comment:
Using a map function and side effects of a list of predicate functions seems inappropriate for this handler.  It is not a good practice to use predicate functions for their side effects as is done here. The documentation in https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg#hdr-Predicate doesn't give any guarantees that predicates with side-effects will work properly.  Here, you are indirectly updating the metrics within the predicate functions, but I'm not clear that this is safe to do.  Can you disentangle the metric updaters from the predicate functions?

Reply:
I filed https://github.com/kubernetes-sigs/controller-runtime/issues/1968 to ask about predicates with side-effects.  Until we get a response on that issue, I share Candace's concerns that using predicates in this way could be problematic.

Reply:
Whenever a ```Watch``` is created on an object a corresponding ```internal.EventHandler``` object is created: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/source/source.go#L158. For every event related to that object the corresponding function (```OnAdd```, ```OnUpdate``` and ```OnDelete```) will be called. The predicates are executed on the object before calling the ```handler.EventHandler``` function.  Create event for example: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/source/internal/eventsource.go#L56-L60. The ```handler.EventHandler``` functions enqueues the corresponding ```reconcile.Request``` of the object in the Queue (eg: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/handler/enqueue.go#L44-L47). The ```reconcile.Request``` is dequeued and processed in the ```processNextWorkItem``` function of the corresponding Controller and the corresponding Reconcile logic is executed: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/controller.go#L255-L275. Thus it can be seen that predicates are part of the flow before the reconcile loop is executed.

There are a few benefits of gathering the metrics in the predicate functions rather than in the reconcile loop.
1. To gather the metrics related to Routes, the event and the corresponding object(s) are required. The predicate functions can directly get this information. To accomplish the same thing in the reconcile loop, first the object has to be fetched from kube-apiserver to figure out what event occurred.
2. There is a possibility of an error occurring during this fetch which will result in re-queuing the object. There's no operation in the current implementation of the predicate functions which will result in an error.
3. Also, for Delete and Update events the old object (the one which was deleted or updated) is needed. This is also directly available in the predicate functions. Additional steps are required to get the same information in the reconcile loop.
4. We also don't need to go through the entire process of enqueuing and dequeuing (also re-queuing) as the same can be accomplished beforehand in the predicate functions with much less complexity.

Reply:
@arkadeepsen, those are implementation details though, not contracts, right?  I agree, using predicates simplifies the logic in some ways and probably improves performance.  However, we need to make sure the logic will continue to work even if controller-runtime changes the implementation details.

Reply:
Ok @Miciah and @candita. I'll change the implementation and remove the dependence on predicate functions.

Reply:
Removed the dependence on predicated functions.